src/scheduler/scheduler.service.ts

```diff
@@ -0,0 +1,83 @@
+import { Injectable } from '@nestjs/common';
+import { CronExpression, SchedulerRegistry } from '@nestjs/schedule';
+import { CronJob } from 'cron';
+import moment from 'moment';
+import { RecurringInterval, Recursion, RecursionPeriod, Schedule } from './interface/scheduler.interface';
+
+@Injectable()
+export class SchedulerService {
+  constructor(private schedulerRegistry: SchedulerRegistry) {}
+  scheduleRecurringJob(schedule: Schedule, callback: () => void) {
+    if (moment(schedule.startDate).utcOffset(0).milliseconds() < Date.now()) throw new RangeError('Invalid Date: date should be after present');
+    const job = this.addRecurringJob(schedule, callback);
+    this.scheduleJob(schedule.name, schedule.startDate, () => {
+      if (schedule.recursion && schedule.recursion.days) {
+        //handle start day doesn't match recurring pattern
+        const cronExpArray = this.getCustomCronExpression(schedule.recursion, schedule.startDate).split(' ');
+        const IsStartDayMatchCron = this.toCsvString(schedule.recursion.days) === cronExpArray[cronExpArray.length - 1];
+        if (IsStartDayMatchCron) callback();
+      } else callback();
+      job.start();
+    });
+  }
+  scheduleJob(jobName: string, startDate: Date, callback: () => void) {
+    if (moment(startDate).utcOffset(0).milliseconds() < Date.now()) throw new RangeError('Invalid Date: date should be after present');
+    const name = `${jobName}-timeout`;
+    const timeDiff = moment(startDate).utcOffset(0).diff(moment(), 'milliseconds');
+    this.addTimeout(name, timeDiff, callback);
+  }
+  private addTimeout(name: string, ms: number, callback: () => void) {
+    const timeout = setTimeout(callback, ms);
+    if (this.schedulerRegistry.getTimeouts().find((timeoutName) => timeoutName === name)) this.schedulerRegistry.deleteTimeout(name);
+    this.schedulerRegistry.addTimeout(name, timeout);
+  }
+  private addRecurringJob({ name, predefined, recursion, startDate }: Schedule, callback: () => void) {
+    const jobId = `${name}-recurring`;
+    if (this.schedulerRegistry.getCronJobs().has(jobId)) this.schedulerRegistry.deleteCronJob(jobId);
```

Comment:
so it means, there is no cronjob that has the same name right since jobId refer to name

Reply:
yes that's correct

Reply:
yes that's correct